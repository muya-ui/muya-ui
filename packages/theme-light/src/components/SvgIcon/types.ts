export interface ISvgPureProps {
  /** svg 元素的 viewBox */
  viewBox?: string;
  /** svg 元素的类名 */
  className?: string;
  /** svg 元素的颜色 */
  color?: string;
  /** svg 元素的字号 */
  fontSize?: string | number;
  /** 子节点 */
  children?: React.ReactElement[] | React.ReactElement;
  /** 图标背景元素的百分比尺寸，宽/高默认为[0.7, 0.7] */
  bgSize?: number[];
  /** 图标背景元素颜色，不传入则不展示背景元素 */
  bgColor?: string;
}

export type ISvgProps = ISvgPureProps & React.SVGAttributes<SVGSVGElement>;
