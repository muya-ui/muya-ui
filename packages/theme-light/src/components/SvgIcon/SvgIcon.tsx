import React, { Ref } from 'react';
import styled from 'styled-components';

import { ISvgProps } from './types';

export const StyledSvg = styled.svg`
  display: inline-block;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  color: ${props => props.color};
  font-size: ${props => props.fontSize};
  fill: currentColor;
`;

export default React.forwardRef<SVGSVGElement, ISvgProps>(function SvgIcon(
  props: ISvgProps,
  ref: Ref<SVGSVGElement>,
) {
  const viewBoxSize = 1024;
  const {
    children,
    className,
    viewBox = `0 0 ${viewBoxSize} ${viewBoxSize}`,
    color,
    bgSize = [],
    bgColor,
    fontSize,
    ...other
  } = props;
  const [bgWidthPercent = 0.7, bgHeightPercent = 0.7] = bgSize;
  const bgWidth = bgWidthPercent * viewBoxSize;
  const bgHeight = bgHeightPercent * viewBoxSize;
  const bgX = (viewBoxSize - bgWidth) / 2;
  const bgY = (viewBoxSize - bgHeight) / 2;
  return (
    <StyledSvg
      focusable="false"
      aria-hidden="false"
      viewBox={viewBox}
      className={className}
      ref={ref}
      color={color}
      fontSize={fontSize}
      {...other}
    >
      <>
        {bgColor && <rect width={bgWidth} height={bgHeight} x={bgX} y={bgY} fill={bgColor} />}
        {children}
      </>
    </StyledSvg>
  );
});
