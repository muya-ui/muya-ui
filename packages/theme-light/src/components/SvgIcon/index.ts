export { default } from './SvgIcon';

export { default as SvgIcon } from './SvgIcon';
export * from './SvgIcon';
export * from './types';
