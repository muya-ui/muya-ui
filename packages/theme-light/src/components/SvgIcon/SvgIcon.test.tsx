import { shallow } from 'enzyme';
import React from 'react';
import renderer from 'react-test-renderer';

import SvgIcon from './index';

test('测试普通渲染', () => {
  const fakeChildren = (
    <>
      <div key="1">1</div>
      <div key="2">2</div>
      <div key="3">3</div>
    </>
  );
  const wrapper = shallow(<SvgIcon>{fakeChildren}</SvgIcon>);
  // receive all children
  expect(wrapper.find('div')).toHaveLength(3);
});

test('测试设置颜色', () => {
  const tree = renderer.create(<SvgIcon color="red" />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('color', 'red');
});

test('测试设置大小', () => {
  const tree = renderer.create(<SvgIcon fontSize="24px" />).toJSON();
  expect(tree).toMatchSnapshot();
  expect(tree).toHaveStyleRule('font-size', '24px');
});

test('测试图标背景元素', () => {
  const wrapper = shallow(<SvgIcon bgColor="#fff"></SvgIcon>);
  // should show bg element
  expect(wrapper.find('rect')).toHaveLength(1);
});
