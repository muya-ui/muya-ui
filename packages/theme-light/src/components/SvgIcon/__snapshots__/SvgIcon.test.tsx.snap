// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`测试设置大小 1`] = `
.c0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  font-size: 24px;
  fill: currentColor;
}

<svg
  aria-hidden="false"
  className="c0"
  focusable="false"
  fontSize="24px"
  viewBox="0 0 1024 1024"
/>
`;

exports[`测试设置颜色 1`] = `
.c0 {
  display: inline-block;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  font-style: normal;
  line-height: 0;
  text-align: center;
  text-transform: none;
  vertical-align: -0.125em;
  text-rendering: optimizeLegibility;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  width: 1em;
  height: 1em;
  overflow: hidden;
  color: red;
  fill: currentColor;
}

<svg
  aria-hidden="false"
  className="c0"
  color="red"
  focusable="false"
  viewBox="0 0 1024 1024"
/>
`;
